File: framework/languages/java/samples/java/Bingo/Server/Play/src/main/java/systems/zlink/samples/bingo/server/play/infrastructure/zlink/actors/PlayerActorTransferAdapter.java
package systems.zlink.samples.bingo.server.play.infrastructure.zlink.actors;

import systems.zlink.framework.actors.ZLinkActorContext;
import systems.zlink.framework.actors.ZLinkActorTransferAdapter;
import systems.zlink.framework.messaging.ZLinkMessage;

public final class PlayerActorTransferAdapter
    implements ZLinkActorTransferAdapter<PlayerActor> {
    @Override
    public java.util.concurrent.CompletionStage<ZLinkMessage> transferOut(
        PlayerActor actor) {
        return java.util.concurrent.CompletableFuture.completedFuture(ZLinkMessage.of(new TransferState(
            actor.displayName(),
            actor.roomId(),
            actor.destroyAfterEntrySpotJoin(),
            actor.disconnected())));
    }

    @Override
    public java.util.concurrent.CompletionStage<PlayerActor> transferIn(
        String actorId,
        ZLinkActorContext context,
        ZLinkMessage state) {
        TransferState transferred = state.decode(TransferState.class);
        PlayerActor actor = new PlayerActor(actorId, context);
        actor.setDisplayName(transferred.displayName());
        if (transferred.roomId() != null && !transferred.roomId().isBlank()) {
            actor.joinRoom(transferred.roomId());
        }
        if (transferred.destroyAfterEntrySpotJoin()) {
            actor.markForDestroyAfterRoomLeave();
        }
        if (transferred.disconnected()) {
            actor.markDisconnected();
        }
        return java.util.concurrent.CompletableFuture.completedFuture(actor);
    }

    public record TransferState(
        String displayName,
        String roomId,
        boolean destroyAfterEntrySpotJoin,
        boolean disconnected) {
    }
}

File: framework/languages/java/samples/kotlin/Bingo/Server/Play/src/main/kotlin/systems/zlink/samples/kotlin/bingo/server/play/infrastructure/zlink/actors/PlayerActorTransferAdapter.kt
package systems.zlink.samples.kotlin.bingo.server.play.infrastructure.zlink.actors

import systems.zlink.framework.actors.ZLinkActorContext
import systems.zlink.framework.kotlin.ZLinkSuspendingActorTransferAdapter
import systems.zlink.framework.messaging.ZLinkMessage

class PlayerActorTransferAdapter : ZLinkSuspendingActorTransferAdapter<PlayerActor>() {
    override suspend fun transferOutSuspending(
        actor: PlayerActor,
    ): ZLinkMessage = ZLinkMessage.of(
        TransferState(
            actor.displayName,
            actor.roomId,
            actor.destroyAfterEntrySpotJoin,
            actor.disconnected,
        ),
    )

    override suspend fun transferInSuspending(
        actorId: String,
        context: ZLinkActorContext,
        state: ZLinkMessage,
    ): PlayerActor {
        val transferred = state.decode(TransferState::class.java)
        return PlayerActor(actorId, context).also { actor ->
            actor.setDisplayName(transferred.displayName)
            if (transferred.roomId.isNotBlank()) actor.joinRoom(transferred.roomId)
            if (transferred.destroyAfterEntrySpotJoin) actor.markForDestroyAfterRoomLeave()
            if (transferred.disconnected) actor.markDisconnected()
        }
    }

    data class TransferState(
        val displayName: String,
        val roomId: String,
        val destroyAfterEntrySpotJoin: Boolean,
        val disconnected: Boolean,
    )
}
